backend/rates/rates.go

```diff
@@ -37,45 +39,103 @@ var fiats = []string{"USD", "EUR", "CHF", "GBP", "JPY", "KRW", "CNY", "RUB", "CA
 const interval = time.Minute
 const cryptoCompareURL = "https://min-api.cryptocompare.com/data/pricemulti?fsyms=%s&tsyms=%s"
 
-// RateUpdater implements coin.RateUpdater.
+type exchangeRate struct {
+	timestamp time.Time
+	value     float64
+}
+
+// RateUpdater provides cryptocurrency-to-fiat conversion rates.
 type RateUpdater struct {
 	observable.Implementation
-	last       map[string]map[string]float64
-	log        *logrus.Entry
-	socksProxy socksproxy.SocksProxy
+
+	httpClient *http.Client
+	log        *log.Logger
+
+	// last contains most recent conversion to fiat, keyed by a coin.
+	last map[string]map[string]float64
+
+	histMu sync.RWMutex
+	// hist contains historical conversion rates in asc order, keyed by a coin.
+	// TODO: store in bolt DB
+	// TODO: support different fiat currencies? it's in USD at the moment
+	hist map[string][]exchangeRate
+
+	// CoinGecko is where updater gets the historical conversion rates.
+	// See https://www.coingecko.com/en/api for details.
+	coingeckoURL string
 }
 
 // NewRateUpdater returns a new rates updater.
-func NewRateUpdater(socksProxy socksproxy.SocksProxy) *RateUpdater {
-	ratesUpdater := &RateUpdater{
-		last:       map[string]map[string]float64{},
-		log:        logging.Get().WithGroup("rates"),
-		socksProxy: socksProxy,
+func NewRateUpdater(client *http.Client) *RateUpdater {
+	return &RateUpdater{
+		last:         make(map[string]map[string]float64),
+		hist:         make(map[string][]exchangeRate),
+		log:          log.New(os.Stderr, "[rates] ", log.LstdFlags|log.Lmsgprefix),
```

Comment:
this now does not go into the log.txt file. i suggest to stick with logrus, and replace all of logrus in a dedicated PR. mixing the two confusing to read and maintain.

Reply:
hm... haven't thought about the log.txt. good point. will revert.

Reply:
reverted